crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2366 @@
+      LN_sha512, 64, 56, 56, 56 }
+};
+
+/*
+ * @brief info about a KDF
+ */
+typedef struct {
+    uint16_t       kdf_id; /**< code point for KDF */
+    const char     *mdname; /**< hash alg name for the HKDF */
+    size_t         Nh; /**< length of hash/extract output */
+} hpke_kdf_info_t;
+
+/*
+ * @brief table of KDFs
+ */
+static hpke_kdf_info_t hpke_kdf_tab[] = {
+    { 0, NULL, 0 }, /* keep indexing correct */
+    { OSSL_HPKE_KDF_ID_HKDF_SHA256, LN_sha256, SHA256_DIGEST_LENGTH },
+    { OSSL_HPKE_KDF_ID_HKDF_SHA384, LN_sha384, SHA384_DIGEST_LENGTH },
+    { OSSL_HPKE_KDF_ID_HKDF_SHA512, LN_sha512, SHA512_DIGEST_LENGTH }
+};
+
+/**
+ * @brief sender or receiver context
+ */
+struct ossl_hpke_ctx_st
+{
+    OSSL_LIB_CTX *libctx; /**< library context */
+    char *propq; /**< properties */
+    int mode; /**< HPKE mode */
+    OSSL_HPKE_SUITE suite; /**< suite */
+    uint64_t seq; /**< sequence number */
+    unsigned char *shared_secret;
+    size_t shared_secretlen;
+    unsigned char *exportersec; /**< exporter secret */
+    size_t exporterseclen;
+    char *pskid; /**< PSK stuff */
+    unsigned char *psk;
+    size_t psklen;
+    EVP_PKEY *senderpriv; /**< sender's ephemeral private key */
+    char *ikme;
+    size_t ikmelen;
+    EVP_PKEY *authpriv; /**< sender's authentication private key */
+    unsigned char *authpub; /**< auth public key */
+    size_t authpublen;
+};
+
+/*
+ * @brief map from IANA codepoint to AEAD table index
+ *
+ * @param codepoint should be an IANA code point
+ * @return index in AEAD table or 0 if error
+ */
+static uint16_t aead_iana2index(uint16_t codepoint)
+{
+    uint16_t naeads = OSSL_NELEM(hpke_aead_tab);
+    uint16_t i = 0;
+
+    for (i = 0; i != naeads; i++) {
+        if (hpke_aead_tab[i].aead_id == codepoint)
+            return i;
+    }
+    return 0;
+}
+
+/*
+ * @brief map from IANA codepoint to KEM table index
+ *
+ * @param codepoint should be an IANA code point
+ * @return index in KEM table or 0 if error
+ */
+static uint16_t kem_iana2index(uint16_t codepoint)
+{
+    uint16_t nkems = OSSL_NELEM(hpke_kem_tab);
+    uint16_t i = 0;
+
+    for (i = 0; i != nkems; i++) {
+        if (hpke_kem_tab[i].kem_id == codepoint)
+            return i;
+    }
+    return 0;
+}
+
+/*
+ * @brief map from IANA codepoint to AEAD table index
+ *
+ * @param codepoint should be an IANA code point
+ * @return index in AEAD table or 0 if error
+ */
+static uint16_t kdf_iana2index(uint16_t codepoint)
+{
+    uint16_t nkdfs = OSSL_NELEM(hpke_kdf_tab);
+    uint16_t i = 0;
+
+    for (i = 0; i != nkdfs; i++) {
+        if (hpke_kdf_tab[i].kdf_id == codepoint)
+            return i;
+    }
+    return 0;
+}
+
+/*
+ * @brief Check if kem_id is ok/known to us
+ * @param kem_id is the externally supplied kem_id
+ * @return 1 for good, not 1 for error
+ */
+static int hpke_kem_id_check(uint16_t kem_id)
+{
+    switch (kem_id) {
+    case OSSL_HPKE_KEM_ID_P256:
```

Comment:
You could also search the hpke_kem_tab[].. If you want to do it this way put a comment above the id's saying what needs to be updated if a new entry is added..

Reply:
think something equivalent will be needed unless/until the difference between nist and cfrg curves could be invisible to callers but not sure if I'm getting what you mean so leaving this open

Reply:
covered by other issue